java-spiffe-core/src/main/java/spiffe/bundle/jwtbundle/JwtBundleSet.java

```diff
@@ -0,0 +1,55 @@
+package spiffe.bundle.jwtbundle;
+
+import lombok.NonNull;
+import lombok.Value;
+import org.apache.commons.lang3.NotImplementedException;
+import spiffe.result.Result;
+import spiffe.spiffeid.TrustDomain;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A <code>JwtBundleSet</code> represents a set of X509Bundles keyed by TrustDomain.
+ */
+@Value
+public class JwtBundleSet implements JwtBundleSource {
+
+    ConcurrentHashMap<TrustDomain, JwtBundle> bundles;
+
+    private JwtBundleSet(ConcurrentHashMap<TrustDomain, JwtBundle> bundles) {
+        this.bundles = bundles;
+    }
+
+    public static JwtBundleSet of(@NonNull final List<JwtBundle> bundles) {
+        throw new NotImplementedException("Not implemented");
+    }
+
+    public static JwtBundleSet of(@NonNull final TrustDomain trustDomain,
+                                  @NonNull final JwtBundle jwtBundle) {
+        throw new NotImplementedException("Not implemented");
+    }
+
+    public List<JwtBundle> getJwtBundles() {
+        return new ArrayList<>(bundles.values());
+    }
+
+    @Override
+    public Result<JwtBundle, String> getJwtBundleForTrustDomain(final TrustDomain trustDomain) {
```

Comment:
Is there any reason to report errors a la Go, instead of using Java exceptions?

Reply:
It's a more _civilized_ way to handle errors, more friendly with the functional style that java is embracing since v8. Instead of throwing exceptions around, you just hand off the caller the result of an operation that can be an Ok containing a Value or an Error containing the information about the error. A la Go is better than throwing exception, IMO.

Reply:
I'm sorry, but I'm not seeing the _civilized_ part of this. IMO Go's error handling is one of its weakest points, and in fact there are efforts to improve this for Go 2. 
I believe that this approach for Java is error prone, first, you have a function that returns one value, which might be the actual object OR an error (in Go the function returns 2 objects, the value AND the error, so it is more explicit about the error), so you will have to end up having to catch unexpected errors anyways. Like in the following function of CertificateUtils.java:
```java
    public static Result<PrivateKey, String> generatePrivateKey(byte[] privateKeyBytes) {
        PKCS8EncodedKeySpec kspec = new PKCS8EncodedKeySpec(privateKeyBytes);
        Result<PrivateKey, Throwable> privateKeyResult = generatePrivateKeyWithSpec(kspec);

        if (privateKeyResult.isOk()) {
            return Result.ok(privateKeyResult.getValue());
        }

        // PrivateKey is in PEM format, not supported, need to convert to DER and try again
        if (privateKeyResult.getError() instanceof InvalidKeySpecException) {
            byte[] keyDer = toDerFormat(privateKeyBytes);
            kspec= new PKCS8EncodedKeySpec(keyDer);
            privateKeyResult = generatePrivateKeyWithSpec(kspec); // <-- This privateKeyResult is not being checked.
        }
        return Result.ok(privateKeyResult.getValue());   // <-- If privateKeyResult is an error, getValue throws a NoSuchElementException.
    }
```
With the regular try...catch I think it should look something like:
```java
    public static PrivateKey generatePrivateKey(byte[] privateKeyBytes) {
        try {
            PKCS8EncodedKeySpec kspec = new PKCS8EncodedKeySpec(privateKeyBytes);
            return generatePrivateKeyWithSpec(kspec);
        } catch (InvalidKeySpecException e) {
            // PrivateKey is in PEM format, not supported, need to convert to DER and try again
            byte[] keyDer = toDerFormat(privateKeyBytes);
            kspec= new PKCS8EncodedKeySpec(keyDer);
            return generatePrivateKeyWithSpec(kspec);
        }
    }
```
Also, IMO any Java developer is used to work with exceptions, and if this library will be used from a pre-existent Java codebase, I believe they must/should already have some proper exception handling.

Disclaimer: I stopped working on Java almost two and a half years ago...so I might be more than rusty :grimacing:. So I believe it would be very useful to have opinions of other Java developers.

Reply:
In this last piece of code in your comment the method _generatePrivateKey_ would need to be marked as throwing NoSuchAlgorithmException and InvalidKeySpecException because  you put _generatePrivateKeyWithSpec_ in the catch, then that catch clause could throw an exception. Thus it wouldn't be so clean after all. The method caller would need to have itself another try-catch to handle those exceptions, etc. 
In java returning two values is not allowed, but using this Result type you can be explicit about the result of an operation conveying _either_  a value or an error, and you don't force the consumer to put a try-catch or think about checked and unchecked exceptions. Thinking about exceptions is more complex, error prone, throwing an exception is like using _goto_, you don't know where it will end up. 
I know that java developers are used to exceptions, but they are used to returning _nulls_ as well and that is a smelly practice, that could be improved with the use of the type _Optional_ added to Java 8. Well this Result type is like an Optional, a wrapper of a value.

Reply:
I would say that despite any rough edges or warts that may come along with the exception handling paradigm, it is at least something that every Java developer understands and expects. And there is something to be said for consistency and operating within well understood language idioms. It results in more readable code which is easier for the average developer to reason about.

Introducing a new error handling paradigm here requires consumers of this library to write their code which interacts with this library in a different way than the rest of their application. I think that in itself presents a barrier to adoption of this library.

I would also add that the usage of `Optional<T>` as a return type of a public method is still a fairly controversial topic within the Java community for various reasons (an `Optional` can still be null, [ it's not serializable](http://mail.openjdk.java.net/pipermail/jdk8-dev/2013-September/003274.html), etc.). I don't think there's strong consensus on the usage of wrapped return types in Java, apart from fluent-style APIs such as `java.util.stream`.

Reply:
To my knowledge it is general consensus in Java (and afaik in any language with exceptions) to use exceptions to communicate exceptional, unexpected states and to use return values for regular control flow. This makes the regular control flow easier to write and read and also enables exception facades. The distinction is of subject to interpretation.

The `Result` / `Option` types introduced with Java 8 are meant as a functional alternative to express regular control flow, not to supersede exceptions.

A typical example for this would be a nullable return value wrapped in an optional, or the result of an HTTP call wrapped in a `Result` object with Non-OK states mapped as error.
If the class `Result` is used, it is good practice to use typed errors or an enum to allow the caller to distinguish error cases.

In the example under discussion it would be appropriate to use exceptions, as invalid keys or algorithms are not "normal" control flow. Also, in Java it would be considered highly unusal to pass around exception objects without them being thrown. Also, `Throwable` should not be used in such a way as this class also covers unrecoverable errors like `OutOfMemoryError`. A tool like SonarQube will point out stuff like that immediately.

Also it is important to take developers expectations into account. For instance for the validate method, a Java developer would likely expect that errors are simply communicated with the `java.security` exceptions, even if a validation failure would be considered regular control flow.

An example for proper use of the Result class would be the `getSpiffeId` method, altough I'd prefer typed errors here.

Reply:
Thanks for your feedback about the error handling using the Result type. I'm working on refactoring it based on your comments.

Reply:
I made the refactor replacing the Result wrapper type by checked and unchecked exceptions.